### ISSUE TITLE
Buffer size of stdin?

### ISSUE THREAD

**Issue by marktheunissen** [fluent/fluent-bit#89]
Hi there, thanks for the cool utility. 

I'm trying to find out how big the buffer size is when reading from stdin. Let's say I have a 10M file and I just want to `cat` that file through fluent bit to my target.

So I do: `cat events.log | td-agent-bit -c td-agent-bit.conf` where the conf file is:

```ini
[SERVICE]
    Flush     1
    Daemon    off
    Log_Level debug

[INPUT]
    Name stdin
    Tag  events

[OUTPUT]
    Name forward
    Match events
    Host example.com
    Port 24224
```

I would expect the entire 10M file to be sent through the pipeline, however the actual result is that I get:

```
TD-Agent Bit v0.8.3
Copyright (C) Treasure Data

[2016/07/25 16:27:12] [ info] starting engine
[2016/07/25 16:27:12] [debug] [router] match rule stdin.0:forward.0
[2016/07/25 16:27:14] [ warn] STDIN data incomplete, waiting for more data...
```

And only a tiny fraction of the file contents are delivered.

Now, I can slow down the pipe using the following:

`cat events.log | pv -l -L 10 -q | td-agent-bit -c td-agent-bit.conf`

But it still doesn't work as I'd expect, I get many more events delivered, but still just about 1/10th. And that's slowing the rate to 10 bytes / second.

Is there a configuration toggle for the size of the buffer? I would like it to keep buffering data up to a certain limit, bounded by the amount of RAM on the box, I guess.

Thanks!

**Comment by marktheunissen** [comment]
I see there is this line in the header:

```c
char buf[8192 * 2];               /* read buffer: 16Kb max */
```

Would changing that help me?

**Comment by edsiper** [comment]
The stdin buffer size is fixed, the workaround would be to increase that.

Anyways I will add more flexible support for that in the next version. Btw,
what kind of data does your file contains?

**Comment by marktheunissen** [comment]
Thanks!

The file is JSON data of events. I'm forwarding them to FluentD running on a cloud VM.

So all I need to do is recompile with a larger buffer size on that exact line in the header, and it will solve the issue? How big can I reasonably go? 100 megs? :)

**Comment by edsiper** [comment]
Increasing the buffer would help a bit but it's not a real solution, let me
get back to you this week with some improvement on that area.

**Comment by nokute78** [comment]
The problem is processing of 'incomplete JSON data'

fluent-bit reported
 `[ warn] STDIN data incomplete, waiting for more data... `

In this case, fluent-bit was waiting following data.
However, ctx->buf was full and ctx->buf_len was also maximum.

Then fluent-bit repeated to read(2) 0bytes. ( because sizeof(ctx->buf) and ctx->buf_len are 16384.)
So, fluent-bit can't get following JSON data.
``` C
    bytes = read(ctx->fd,
                 ctx->buf + ctx->buf_len,
                 sizeof(ctx->buf) - ctx->buf_len);
```

**Comment by nokute78** [comment]
To fix it, we need to know position which is end of JSON data.

for example, fluent-bit read such data.
`{"foo":"bar"}{"foo":"bar"}{"foo":"bar"}...{"foo":"bar"}{"foo":"bar"}{"hoge":`

fluent-bit should pack 
`{"foo":"bar"}{"foo":"bar"}{"foo":"bar"}...{"foo":"bar"}{"foo":"bar"}`

and 
 should set `ctx->buf` to the rest of JSON `{"hoge":`  and continue to read following data.

**Comment by marktheunissen** [comment]
What about using a newline delimiter? That's how my data is structured:

```
{"foo":"bar"}\n
{"foo":"bar"}\n
{"foo":"bar"}\n
```

**Comment by nokute78** [comment]
@marktheunissen 
I also tested with a data using \n delimiter. (Oops I forgot to write a delimiter above comment.)

Could you test this script for workaround ?

``` sh
#!/bin/sh

while read line 
do
  echo ${line}
  usleep 30000
done < $1
```

How to use.
```
$ chmod +x workaround.sh
$ workaround.sh event.log | td-agent-bit -c td-agent-bit.conf
```

**Comment by edsiper** [comment]
@marktheunissen can you upload an example file with the JSON samples ? (100MB ideal)

**Comment by edsiper** [comment]
I have pushed a new branch called [stdin_buffer](https://github.com/fluent/fluent-bit/tree/stdin_buffer), this one adds the following features:

- support concatenated JSON maps
- incoming buffer now is dynamic, by default is set to 32KB
- new configuration property _buffer_size_ to adjust the incoming buffer. The value must answer the following question: does __each__ JSON map fit in 32KB ?, if not, use a different value.

Please get a copy of that branch, build it and give it a try with:

```bash
$ cat test.json | fluent-bit -i stdin -t test -o stdout -m test 
```
please test and let me know how it works.

**Comment by marktheunissen** [comment]
Cool, I tested and it works, thanks!

`cat events.log | /usr/bin/stdbuf -oL ~/src/fluent-bit/build/bin/fluent-bit -i stdin -o stdout -f 1 > outtest`

now `cat outtest | wc -l` gives the right number of lines 👍

**Comment by marktheunissen** [comment]
I have noticed another issue, but I'll open a new ticket.

**Comment by edsiper** [comment]
thanks, I will merge this new feature in 0.8 branch.

**Comment by nokute78** [comment]
@edsiper 
I also tested [stdin_buffer](../tree/stdin_buffer), then segfault occurred.
`ctx->buf_len` was increased more than `ctx->buf_size`
Could you check it ?
Should we also reset `ctx->buf_len` in case of `ret == FLB_ERR_JSON_INVAL` ?

I generated JSON data using this script.
```sh
#!/bin/sh

for i in `seq 0 10000`
do
    echo "{\"foo\":\"bar\"}"
done
```
```
$ sh output.sh > data.txt
$ cat data.txt | bin/fluent-bit -i stdin -o stdout  -vv
```

## result

```
$ cat data.txt |bin/fluent-bit -i stdin -o stdout -f 1 -vv
Fluent-Bit v0.8.4
Copyright (C) Treasure Data

[2016/07/30 15:33:57] [ info] starting engine
[2016/07/30 15:33:57] [debug] [stdin] buffer_size=32768 bytes
[2016/07/30 15:33:57] [debug] [router] default match rule stdin.0:stdout.0
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
[2016/07/30 15:33:57] [debug] [in_serial] invalid JSON message, skipping
Segmentation fault
```

**Comment by edsiper** [comment]
thanks for catching that problem. I fixed the INVAL reset stuff here ad187233cfd46045e38562b998c2fead520733b9

note: I merged all changes in 0.8 branch (stdin_buffer should not longer be used).

**Comment by nokute78** [comment]
@edsiper Thanks. I will check it.
